crates/core/tedge_agent/src/http_rest.rs

```diff
@@ -0,0 +1,371 @@
+use futures::StreamExt;
+use hyper::{server::conn::AddrIncoming, Body, Request, Response, Server};
+use path_clean::PathClean;
+use routerify::{Router, RouterService};
+use std::path::Path;
+use std::{net::IpAddr, net::SocketAddr, path::PathBuf};
+
+use tedge_utils::paths::create_directories;
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+
+use crate::error::FileTransferError;
+
+#[derive(Debug, Clone)]
+pub struct HttpConfig {
+    pub bind_address: SocketAddr,
+    pub file_transfer_uri: String,
+    pub file_transfer_dir: PathBuf,
+}
+
+impl Default for HttpConfig {
+    fn default() -> Self {
+        HttpConfig {
+            bind_address: ([127, 0, 0, 1], 3000).into(),
+            file_transfer_uri: "/tedge/".into(),
+            file_transfer_dir: "/var/tedge/".into(),
+        }
+    }
+}
+
+impl HttpConfig {
+    pub fn with_ip_address(self, ip_address: IpAddr) -> HttpConfig {
+        Self {
+            bind_address: SocketAddr::new(ip_address, self.bind_address.port()),
+            ..self
+        }
+    }
+
+    #[cfg(test)]
+    pub fn with_file_transfer_dir(self, file_transfer_dir: PathBuf) -> HttpConfig {
+        Self {
+            file_transfer_dir,
+            ..self
+        }
+    }
+
+    pub fn file_transfer_end_point(&self) -> String {
+        format!("{}*", self.file_transfer_uri)
+    }
+
+    pub fn file_transfer_dir_as_string(&self) -> String {
+        self.file_transfer_dir
+            .to_str()
+            .expect("Non UTF8 http root path")
+            .into()
+    }
+
+    /// Return the path of the file associated to the given `uri`
+    ///
+    /// Check that:
+    /// * the `uri` is related to the file-transfer i.e a sub-uri of `self.file_transfer_uri`
+    /// * the `path`, once normalized, is actually under `self.file_transfer_dir`
+    pub fn local_path_for_uri(&self, uri: String) -> Result<PathBuf, FileTransferError> {
+        let ref_uri = uri.clone();
+
+        // The file transfer prefix has to be removed from the uri to get the target path
+        let path = uri
+            .strip_prefix(&self.file_transfer_uri)
+            .ok_or(FileTransferError::InvalidURI { value: ref_uri })?;
+
+        // This path is relative to the file transfer dir
+        let full_path = self.file_transfer_dir.join(path);
+
+        // One must check that once normalized (i.e. any `..` removed)
+        // the path is still under the file transfer dir
+        let clean_path = full_path.clean();
+        if clean_path.starts_with(&self.file_transfer_dir) {
+            Ok(clean_path)
+        } else {
+            Err(FileTransferError::InvalidURI {
+                value: clean_path.to_string_lossy().to_string(),
+            })
+        }
+    }
+}
+
+mod uri_utils {
```

Comment:
Any specific reason why you created a dedicated module here rather thane keeping that function at the root level? I'm not arguing against it. Just trying to understand what was special about this simple utility function.

Reply:
My bad! I removed all but one function from this module. The module can be removed. The function could also be made clearer.